core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HealthCheckedEndpointGroup.java

```diff
@@ -187,40 +214,175 @@ public MeterBinder newMeterBinder(MeterIdPrefix idPrefix) {
+            executor = new PerEndpointExecutor(clientFactory.eventLoopGroup());
         }
 
-        Endpoint endpoint() {
+        void init(AsyncCloseable handle) {
+            assert this.handle == null;
+            this.handle = handle;
+        }
+
+        @Override
+        public Endpoint endpoint() {
             return endpoint;
         }
+
+        @Override
+        public ClientFactory clientFactory() {
+            return clientFactory;
+        }
+
+        @Override
+        public SessionProtocol protocol() {
+            return protocol;
+        }
+
+        @Override
+        public int port() {
+            return port;
+        }
+
+        @Override
+        public Function<? super ClientOptionsBuilder, ClientOptionsBuilder> clientConfigurator() {
+            return clientConfigurator;
+        }
+
+        @Override
+        public ScheduledExecutorService executor() {
+            return executor;
+        }
+
+        @Override
+        public long nextDelayMillis() {
+            final long delayMillis = retryBackoff.nextDelayMillis(1);
+            if (delayMillis < 0) {
+                logger.warn("Health check might have stopped working due to a negative delayMillis ({}) " +
+                            "returned by retryBackOff.nextDelayMillis(0).", delayMillis);
+                return 0;
+            }
+
+            return delayMillis;
+        }
+
+        @Override
+        public void updateHealth(double health) {
+            final boolean updated;
+            if (health > 0) {
+                updated = healthyEndpoints.add(endpoint);
+            } else {
+                updated = healthyEndpoints.remove(endpoint);
+            }
+
+            if (updated) {
+                // Rebuild the endpoint list and notify.
+                setEndpoints(delegate.endpoints().stream()
+                                     .filter(healthyEndpoints::contains)
+                                     .collect(toImmutableList()));
+            }
+
+            initialCheckFuture.complete(null);
+        }
+
+        CompletableFuture<?> destroy() {
+            assert handle != null : handle;
+            return handle.closeAsync().handle((unused1, unused2) -> {
+                executor.cancelScheduledFutures();
+                initialCheckFuture.complete(null);
+                return null;
+            });
+        }
+    }
+
+    private static final class PerEndpointExecutor
+            extends AbstractExecutorService implements ScheduledExecutorService {
+
+        private final EventExecutorGroup delegate;
+        private final Set<Future<?>> scheduledFutures =
+                Collections.newSetFromMap(new NonBlockingIdentityHashMap<>());
+
+        PerEndpointExecutor(EventExecutorGroup delegate) {
+            this.delegate = delegate;
+        }
+
+        void cancelScheduledFutures() {
+            for (final Iterator<Future<?>> i = scheduledFutures.iterator(); i.hasNext();) {
```

Comment:
Just want to confirm this is thread-safe, I'm not sure off the top of head.

Reply:
`scheduledFutures` itself is thread-safe, but I guess there's a race condition like this:

1. A health checker decides to schedule a task.
2. `cancelScheduledFutures()` is called by `context.destroy()`.
3. The health checker at (1) schedules a task, which will never be cancelled.

Reply:
Fixed. Not sure I like how I fixed it though. Let me know if you have a better idea.